src/protocol/sender.rs

```diff
@@ -0,0 +1,622 @@
+    first_seq: SeqNumber,
+}
+
+impl SendBuffer {
+    pub fn new(settings: &ConnectionSettings) -> Self {
+        Self {
+            buffer: Default::default(),
+            first_seq: settings.init_seq_num,
+        }
+    }
+
+    pub fn release_acknowledged_packets(&mut self, acknowledged: SeqNumber) {
+        while acknowledged > self.first_seq {
+            self.buffer.pop_front();
+            self.first_seq += 1;
+        }
+    }
+
+    pub fn get<'a, I: Iterator<Item = SeqNumber> + 'a>(
+        &'a self,
+        numbers: I,
+    ) -> impl Iterator<Item = Result<&'a DataPacket, SeqNumber>> + 'a {
+        numbers.map(
+            move |number| match self.buffer.get((number - self.first_seq) as usize) {
+                Some(p) => Ok(p),
+                None => Err(number),
+            },
+        )
+    }
+
+    pub fn push_back(&mut self, data: DataPacket) {
+        self.buffer.push_back(data);
+    }
+}
+
+pub struct LossList {
+    pub list: VecDeque<DataPacket>,
+}
+
+impl LossList {
+    pub fn new(_settings: &ConnectionSettings) -> Self {
+        Self {
+            list: VecDeque::new(),
+        }
+    }
+
+    pub fn push_back(&mut self, packet: DataPacket) {
+        self.list.push_back(packet);
+    }
+
+    pub fn pop_front(&mut self) -> Option<DataPacket> {
+        self.list.pop_front()
+    }
+
+    pub fn remoeve_acknowledged_packets(&mut self, acknowledged: SeqNumber) -> u32 {
+        let mut retransmited_packets = 0;
+        while let Some(x) = self.list.front() {
+            if acknowledged > x.seq_number {
+                let _ = self.pop_front();
+                // this means a packet was lost then retransmitted
+                retransmited_packets += 1;
+            } else {
+                break;
+            }
+        }
+        retransmited_packets
+    }
+
+    pub fn back(&self) -> Option<&DataPacket> {
+        self.list.back()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.list.is_empty()
+    }
+}
+
+#[derive(Debug)]
+pub enum SenderError {}
+
+pub type SenderResult = Result<Option<(ControlPacket, SocketAddr)>, SenderError>;
+
+pub struct SenderMetrics {
+    /// Round trip time, in microseconds
+    pub rtt: i32,
+
+    /// Round trip time variance
+    pub rtt_var: i32,
+
+    /// packet arrival rate
+    pub pkt_arr_rate: i32,
+
+    /// estimated link capacity
+    pub est_link_cap: i32,
+
+    /// Total lost packets
+    pub lost_packets: u32,
+
+    /// Total retransmitted packets
+    pub retrans_packets: u32,
+
+    /// Total received packets (packets that have been ACKed)
+    pub recvd_packets: u32,
+}
+
+impl SenderMetrics {
+    pub fn new() -> Self {
+        Self {
+            rtt: 10_000,
+            rtt_var: 0,
+            pkt_arr_rate: 0,
+            est_link_cap: 0,
+            lost_packets: 0,
+            retrans_packets: 0,
+            recvd_packets: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum SenderAlgorithmStep {
```

Comment:
A comment with a link to the spec would be good here, so it can be deciphered what each step means. Alternatively, you could come up with descriptive names and leave the step numbers in the comments

Also, a comment with what the `Instants` mean might be good